### DISCUSSION TITLE
Type hinting for foreign key fields

### DISCUSSION THREAD

**Discussion by bencrouse** [tortoise/tortoise-orm#1585] [Q&A]
Hey all, I'm setting up a model like this

```python
class User(Model):
    id = fields.UUIDField(pk=True)
    email = fields.CharField(max_length=255, unique=True, index=True)
    organization: fields.ForeignKeyRelation[Organization] = fields.ForeignKeyField(
        "user.Organization", related_name="users"
    )
```

It autogenerates the `user.organization_id` on the model and it works great, but when I run mypy type hinting, it complains the field doesn't exist with an error

```
app/routers/dependencies.py:150: error: "User" has no attribute "organization_id"; maybe "organization"?  [attr-defined]
```

I realize I can fetch the organization model, but this is inefficient, I don't want to run the extra query and/or `JOIN` for that. Is there a different way to define this such that field is explicitly declared? I can add a `@property` for it or something, but I thought I'd check if there was a nicer way to handle it.

Thanks for the great library!

**Comment by vlakius** [comment]
Take a look here : https://tortoise.github.io/models.html?h=type#improving-relational-type-hinting

in your example i think you can fix it in this way:
```python
class User(Model):
    id = fields.UUIDField(pk=True)
    email = fields.CharField(max_length=255, unique=True, index=True)
    organization: fields.ForeignKeyRelation[Organization] = fields.ForeignKeyField(
        "user.Organization", related_name="users"
    )
   # for type hinting
    organization_id: fields.ForeignKeyRelation[Organization]
  


```

**Reply by bencrouse** [reply]
Yeah, I tried that and it worked fine until I tried migrations iirc (aerich raised some error I can't quite remember).

I ended up just doing something like `organization_id: Annotated[UUID, "for type hinting"]` and that fixed mypy.

It does seem something worth resolving to me, any sense of if it's worth opening an issue/PR for if I dig up the error?

**Reply by vlakius** [reply]
IMHO it is worth exploring and I think contributions are always welcome.

**Reply by abondar** [reply]
Would be good to improve type hinting
If you are also able to provide more info about error in aerich - I think it could be useful too, maybe @long2ice could look into fixing it too
